---
layout: post
title:  "Sequence-based ab initio prediction of variant effects on expression and disease risk using ExPecto"
date: 2019-07-04
comments: True
mathjax: True
---

<h2>Introduction:</h2>
Human DNA sequence consist of genes and regulatory sequences. Genes are used for making of proteins while regulatory sequences which is also called as non-coding variant act as a regulatory body which decides many genomic functions like when protein should be made, in what quantity it should be made. Today a biggest challenge in human genetics, precision medicine and evolutionary biology is deciphering the regulatory code of gene expression, including understanding the transcriptional effects of genome variation. To cope up with this problem,we used deep learning based framework known as ExPecto. ExPecto can predict ab initio from  DNA sequence, the tissue specific transcriptional effects of mutations, including rare or never observed.

<h2>ExPecto:</h2>
ExPecto model takes DNA sequence as input and predict the risk of disease which a person can have in future. It uses a deep learning network model follwed by spatial transformation and finally it makes tissue-specific prediction using L2-regularized linear regression models fitted by gradient boosting algorithm. Thus after getting tissue specific prediction, we infer the risk of disease which patient can have.

<h2>Model Architecture:</h2>
{% include image.html align="center" url="/assets/img/nihms965313f1.jpg" description="Deep learning-based sequence model accurately predicts cell type-specific gene expression" %}

<h2>Input:</h2>
ExPecto takes DNA sequence of length 40,000. Basically we find the position of allele and then takes DNA sequence of length 20,000 on it's left and right side. If we take size smaller than 40k, the performance decreases while if we inrease the size more than 40k, there is negligible increase in performance. It can be seen in below figure.  
{% include image.html align="center" url="/assets/img/gen.png" description="Comparison of ExPecto eQTL prediction performance across models trained with different sequence window sizes." %}

In above figure, each line shows performance for the model trained with different window sizes (+/- around the TSS). Â X-axis represents the predicted effect magnitude cutoff, as measured by absolute log fold-change. Y-axis represents the accuracy of predicting the expression change directionality for the variants above the corresponding effect magnitude.

<h2>Architecture:</h2>
ExPecto model consist of three components that act sequentially.<br/>

<h4>1) The deep convolutional network regulatory feature representation model architecture:</h4>
In total, we have 2002 genome-wide histone marks, transcription factor binding and chromatin accessibility features. Our deep learning model scans the sequences each of length 2000. The deep convolutional neural network model predicts epigenomic features of a 200bp region, while also using the 1800bp surrounding context sequence. Thus each 200 bp region will give 2002 features, thus in total we will have 400400 features.
Following is the architecture of deep convolutional network:<br/>
Input (Size: 4bases x 2000bp) =><br/>
    (Layer 1): Convolution(4 -> 320, kernel size=8)<br/>
    (Layer 2): ReLU<br/>
    (Layer 3): Convolution(320 -> 320, kernel size=8)<br/>
    (Layer 4): ReLU<br/>
    (Layer 5): Dropout(Probability=0.2)<br/>
    (Layer 6): Max pooling(pooling size=4)<br/>
    (Layer 7): Convolution(320 -> 480, kernel size=8)<br/>
    (Layer 8): ReLU<br/>
    (Layer 9): Convolution(480 -> 480, kernel size=8)<br/>
    (Layer 10): ReLU<br/>
    (Layer 11): Dropout(Probability=0.2)<br/>
    (Layer 12): Max pooling(pooling size=4)<br/>
    (Layer 13): Convolution(480 -> 960, kernel size=8)<br/>
    (Layer 14): ReLU<br/>
    (Layer 15): Convolution(960 -> 960, kernel size=8)<br/>
    (Layer 16): ReLU<br/>
    (Layer 17): Dropout(Probability=0.2)<br/>
    (Layer 18): Linear(101760 -> 2003)<br/>
    (Layer 19): ReLU<br/>
    (Layer 20): Linear(2003 -> 2002)<br/>
    (Layer 21): Sigmoid<br/>
    => Output (Size: 2002 epigenomic features)<br/>

ReLU indicates the rectified linear unit activation function. Sigmoid indicates the Sigmoid activation function.  Notations such as 4 -> 320 indicate the input and output channel size for each layer.

<h4> 2) Spatial Feature Transformation: </h4> 
The second component of ExPecto is the spatial transformation module that reduces the dimensionality of the learning problem by generating spatially-transformed features.The spatial transformation module reduces the input dimensionality with ten exponential functions weighting upstream and downstream regions separately, with weights based on relative distance to TSS (transformed features with higher decay rate are more concentrated near TSSs). This effectively reduces the number of features 20 fold to 20020. The exponential functions were prespecified (based on empirical selection) and not learned during training.
Scanning the 40kb region surrounding the TSS with the deep convolutional network model outputs 200 spatial bins for each of the 2002 histone mark, transcription factor, and DNase features. Spatial transformation condenses this information into 10 spatial features by using exponential functions with 5 different decay rates to compute weighted sum over spatial bins. Upstream regions and downstream regions are transformed independently.
{% include image.html align="center" url="/assets/img/gen1.png" description="Schematic representation of spatial transformation in ExPecto. " %}


<h4> 3) L2-regularized linear regression model: </h4>
Finally, to make tissue-specific expression predictions, spatially-transformed features are used to predict gene expression levels for each tissue with L2-regularized linear regression models fitted by gradient boosting algorithm. The expression is given by following formula.

$$ expression = \sum_{d \in D}\sum_{i} p_{id}  \Bigg[ \sum_{k} 1(t_{d} < 0) \beta^{up}_{ik} e^{-a_{k} \left\lfloor\dfrac{\abs{t_{d}}}{200bp}\right\rfloor} + \sum_{k} 1(t_{d} > 0) \beta^{down}_{ik} e^{-a_{k} \left\lfloor\dfrac{\abs{t_{d}}}{200bp}\right\rfloor} \Bigg] $$


<h2>Summary:</h2>
So using this model we can predict the risk of disease in a person and person can start the treatment so that it should not occur.
